source/develop/release-management/features/storage/4k-blocks.md

```diff
@@ -0,0 +1,93 @@
+---
+title: 4k block device support for the storage domains
+category: feature
+authors: dchaplyg
+feature_name: 4k block device support for the storage domains
+feature_modules: vdsm
+feature_status: Planning
+---
+
+# 4k block device support for the storage domains
+
+# Summary
+
+VDSM assumes, that 512b is a minimally accessible block on the storage device, which can be written or read atomically. That was true for a long time, but now most of the storage devices changed their block size to the 4096 bytes (4 kilobytes). It is still compatible with VDSM, case while we are requesting 512 atomic read or write, underlying storage will atomically read 4096 bytes, throw away 3584 bytes and return just 512. As it still acts same way, it has no impact on the VDSM. 
+
+Unfortunately, some 4k block storage devices either refuse to read/write smaller blocks atomically (some SAN appliances) or have a performance impact while emulating 512b block device (like VDO). 
+
+Therefore without 4k block device support oVirt either can't use those devices as a storage domain or will have performance degradation on those domains.
+
+## Owner
+
+*   Name: [Denis Chaplygin](https://github.com/akashihi)
+*   Email: <dchaplyg@redhat.com>
+
+## Current status
+
+*   Target Release: 4.3
+*   Status: Planning
+*   Last updated: July 31, 2018
+
+# Implementation details
+
+VDSM heavily relies on 512 block support and it is hardcoded to a lots of places.
+
+The general approach is to refactor all block size dependent code to use block size from the storage domain, while keeping it hardcoded to 512b in it and in a second step, implement block size detection at the SD level and replace hardcoded value with actual value.
+
+Subsystem specific notes are below:
+
+## Sanlock
+
+Sanlock is used by oVirt for locking and fencing shared resources. Latest released sanlock have native support for 4k devices, but, unfortunately, at the moment it work only with block devices, not with filesystems. Speaking precisely, sanlock on filesystems always assumes, that block size of the device is 512b. With patch [1] it will be possible, to specify desired block size, but it is still not able to detect block size of the filesystem backing device.
+
+## Blocksize detection
+
+VDSM have to detect block size during storage domain creation. For block device backed storage domains sanlock or blkid could be used. For filesystem backed storage domains VDSM can actually try to make 512b atomic operation and switch to 4k only if it fails. Something like dd if=/fs/under/test/file of=/dev/null bs=512 count=1 iflag=direct
+
+## Metadata
+
+VDSM needs to know size of block of the Storage domain and the best way to store it is metadata. Metadata for block device backed domains already includes corresponding fields:
+
+* DMDK_LOGBLKSIZE
+* DMDK_PHYBLKSIZE
```

Comment:
I don't think we need the physical block size. We need only the logical block size used for alignment when using direct I/O.

Maybe the physical block size is used in lvm, we reject adding a PV with different block size, not sure if we check the logical or physical block size in this case.

For file storage we probably use only DMKD_LOGBLKSIZE, but I prefer to use something more readable like BLOCK_SIZE. Anyway we can handle the details later, lets not mention it here.

The only think that matters at this point is:
- Detect underlying block size
- Keep block size in storage domain metadata

Reply:
I would like to have same metadata for both block domains and same domains, so we will have same code for both.

In case PHYBLKSIZE isn't used anywhere, we can drop it safely.

Reply:
We don't want to change block storage in any way, this is risky. So we should have new options for file storage. If we add support for 4k to block storage in the future, we will probably replace the old names with the new names.